apps/prairielearn/elements/pl-big-o-input/pl-big-o-input.py

```diff
@@ -112,6 +122,45 @@ def render(element_html: str, data: pl.QuestionData) -> str:
 
     score = data["partial_scores"].get(name, {}).get("score")
 
+    # Prepare parse error
+    parse_error: Optional[str] = data["format_errors"].get(name)
+    missing_input = False
+    feedback = None
+    a_sub = None
+    a_sub_raw = None
+    raw_submitted_answer = None
+
+    if parse_error is None and name in data["submitted_answers"]:
+        a_sub = sympy.latex(
+            phs.convert_string_to_sympy(
+                data["submitted_answers"][name],
+                variables,
+                allow_complex=False,
+                allow_trig_functions=False,
+            )
+        )
+        a_sub_raw = data["raw_submitted_answers"].get(name, None)
+
+        if name in data["partial_scores"]:
+            feedback = data["partial_scores"][name].get("feedback")
+    elif name not in data["submitted_answers"]:
+        missing_input = True
+        parse_error = None
+    else:
+        # Use the existing format text in the invalid popup.
+        with open(BIG_O_INPUT_MUSTACHE_TEMPLATE_NAME, "r", encoding="utf-8") as f:
+            info = chevron.render(f, info_params).strip()
+
+        # Render invalid popup
+        raw_submitted_answer = data["raw_submitted_answers"].get(name)
+        if isinstance(parse_error, str):
+            with open(BIG_O_INPUT_MUSTACHE_TEMPLATE_NAME, "r", encoding="utf-8") as f:
+                parse_error += chevron.render(
+                    f, {"format_error": True, "format_string": info}
+                ).strip()
+        if raw_submitted_answer is not None:
+            raw_submitted_answer = pl.escape_unicode_string(raw_submitted_answer)
```

Comment:
Nit: I don't understand why the `if isinstance(...)` block is sandwiched between the code dealing with `raw_submitted_answer`. Can we keep the `raw_submitted_answer is not None` bit adjacent to the initialization of `raw_submitted_answer` itself?

Reply:
Reading further into the function, I also don't understand why we're reading `raw_submitted_answer` this high in the scope anyways. It seems like the code for the `question` panel reads `raw_submitted_answer` again anyways?

Reply:
You raise some good points actually. I think there are actually a couple of things about this block that can be simplified. I'll make a pass and re-request a review once I can cut down some of this logic.

Reply:
Made a push simplifying a lot of this logic, should be much easier to read now.